File: 17_subsets/subsets.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/subsets
@Language: Python
@Datetime: 16-05-08 22:42
'''

class Solution:
    """
    @param S: The set of numbers.
    @return: A list of lists. See example.
    """
    def subsets(self, S):
        # write your code here
        self.result = []
        if S is None or len(S) == 0:
            return [[]]
        self.search(sorted(S), [], 0)
        return self.result
    
    def search(self, S, subset, pos):
        if pos == len(S):
            self.result.append(subset)
            return
        self.search(S, subset+[S[pos]], pos+1)
        self.search(S, subset, pos+1)
            

File: 17_subsets/subsets.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/subsets
@Language: Java
@Datetime: 16-05-08 22:32
*/

class Solution {
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsets(int[] nums) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<>();
        int n = nums.length;
        Arrays.sort(nums);
        for (int i = 0; i < (1 << n); i++) {
            ArrayList<Integer> subset = new ArrayList<Integer>();
            for (int j = 0; j < n; j++) {
                // check whether the jth digit in i's binary representation is 1
                if ((i & (1 << j)) != 0) {
                    subset.add(nums[j]);
                }
            }
            result.add(subset);
        }
        return result;
    }
    
}